x-pack/legacy/plugins/file_upload/server/routes/file_upload.js

```diff
@@ -5,43 +5,126 @@
+        }
+      )
+    ),
+  },
+  { allowUnknowns: true }
+);
+
+const options = {
+  body: {
+    maxBytes: MAX_BYTES,
+    accepts: ['application/json'],
+  },
+};
+
+export const idConditionalValidation = (body, boolHasId) =>
+  schema
+    .object(
+      {
+        data: boolHasId
+          ? schema.arrayOf(schema.object({}, { allowUnknowns: true }), { minSize: 1 })
+          : schema.any(),
+        settings: boolHasId
+          ? schema.any()
+          : schema.object(
+              {},
+              {
+                defaultValue: {
+                  number_of_shards: 1,
+                },
+                allowUnknowns: true,
+              }
+            ),
+        mappings: boolHasId
+          ? schema.any()
+          : schema.object(
+              {},
+              {
+                defaultValue: {},
+                allowUnknowns: true,
+              }
+            ),
+      },
+      { allowUnknowns: true }
+    )
+    .validate(body);
 
-    const requestContentWithDefaults = {
-      id,
-      callWithRequest: callWithRequestFactory(elasticsearchPlugin, requestObj),
-      index: undefined,
-      settings: {},
-      mappings: {},
-      ingestPipeline: {},
-      data: [],
-      ...requestObj.payload,
-    };
-    return importData(requestContentWithDefaults).catch(wrapError);
+const finishValidationAndProcessReq = (elasticsearchPlugin, getSavedObjectsRepository) => {
+  return async (con, req, { ok, badRequest }) => {
+    const {
+      query: { id },
+      body,
+    } = req;
+    const boolHasId = !!id;
+
+    let resp;
+    try {
+      const validIdReqData = idConditionalValidation(body, boolHasId);
+      const callWithRequest = callWithRequestFactory(elasticsearchPlugin, validIdReqData);
+      const { importData: importDataFunc } = importDataProvider(callWithRequest);
+
+      const { index, settings, mappings, ingestPipeline, data } = validIdReqData;
+      const processedReq = await importDataFunc(
+        id,
+        index,
+        settings,
+        mappings,
+        ingestPipeline,
+        data
+      );
+
+      if (processedReq.success) {
+        resp = ok({ body: processedReq });
+        // If no id's been established then this is a new index, update telemetry
+        if (!boolHasId) {
+          await updateTelemetry({ elasticsearchPlugin, getSavedObjectsRepository });
+        }
+      } else {
+        resp = badRequest(`Error processing request 1: ${processedReq.error.message}`, ['body']);
+      }
+    } catch (e) {
+      resp = badRequest(`Error processing request 2: : ${e.message}`, ['body']);
+    }
+    return resp;
   };
-}
+};
+
+export const initRoutes = (router, esPlugin, getSavedObjectsRepository) => {
+  router.post(
+    {
+      path: `${IMPORT_ROUTE}{id?}`,
+      validate: {
+        query: querySchema,
+        body: bodySchema,
+      },
+      options,
+    },
+    finishValidationAndProcessReq(esPlugin, getSavedObjectsRepository)
```

Comment:
I believe this wrapper is unnecessary. The `context` argument (1st arg) to your route handler should already have what you need under `context.core.elasticsearch` and `context.core.savedObjects`

Reply:
You're likely right on this, but I'm going to address these 2 dependencies in a follow-on PR since I think those may be some of the final dependencies we need to resolve to move this plugin. I originally did migrate them in a previous PR and had to back out the changes due to some issues